platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/ValueState.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.Action;
+import org.gradle.internal.Cast;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.logging.text.TreeFormatter;
+import org.gradle.internal.state.ModelObject;
+
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+/**
+ * Manages values that are finalizable and support conventions.
+ *
+ * @param <S>
+ */
+public abstract class ValueState<S> {
```

Comment:
This used to be `AbstractionProperty.FinalizationState`, with a few minor additions